### ISSUE TITLE
New functions?

### ISSUE THREAD

**Issue by zbozo** [arkenthera/cef3d#3]
Hello, i noticed that your branch develop has some functions but I did not see any use of them, could you tell me what it does?

Cef3DBrowser::CreateJsObject

thank you

**Comment by 0x61726b** [comment]
Hi, this project is pretty much dead. I've added C++ -> javascript integration to use in another project and that was pretty much it. It sets a javascript object from C++ side so you can access the variables in javascript.

**Comment by zbozo** [comment]
is a great project, awesomium has been unavailable for a long time, I ended up finding your project, apparently it is very light because it is an old version of cef, it is good for games
I'm going to use it, how should I call the variables in js for

Cef3DJsValue test("test", "my string");
this->CreateJsObject(test);

**Comment by 0x61726b** [comment]
That would be `window.test` in javascript. Here is a sample project where i did use those stuff.

https://github.com/arkenthera/cef3d/tree/develop/Demos/MMD_Nowplaying

**Comment by zbozo** [comment]
It's what I tested, it's giving undefined, I thought it would be another way :(

**Comment by 0x61726b** [comment]
When are you calling `CreateJsObject` ? Can you post your full code?

**Comment by zbozo** [comment]
Cef3DRenderer::OnProcessMessageReceived
http://prntscr.com/ko9ysr
(i received all msgbox)

SampleBrowser::OnAfterCreated
http://prntscr.com/koa1j0
(i received the messagebox)

My html code:

```html
<!DOCTYPE html>
<html>
<head>
    <script>
        window.addEventListener('load', function() {
            var myelement = document.getElementById("xx");
            myelement.innerHTML = window.test;
        })
    </script>
</head>
<div id="xx">???</div>
</body>
</html>

```

Is there any way to send javascript values to the client like in awesomium?
http://prntscr.com/koa7va

The only way I found it was using V8::Execute in the subprocess, however, it does not have access to client variables to return the requested values in javascript
It would also be a more dynamic form because whenever requested will take the current value of the variable instead of the fixed value

**Comment by 0x61726b** [comment]
Yes, though that is a little more complicated. In the video below i was using a color picker from html side to send color values to c++. I will do a write-up when i have time(sometime today) about how you can do that.

https://youtu.be/CvzLLPfRe34

**Comment by zbozo** [comment]
oh ok, thank you very much :P

**Comment by 0x61726b** [comment]
So, here I will try to explain but bear with me since its been few years since I last worked on this.

To be able to call C++ code from Javascript, first you need an ["extension"](https://bitbucket.org/chromiumembedded/cef/wiki/JavaScriptIntegration.md#markdown-header-extensions). 

The extension is a javascript code which will be uploaded by renderer process. Internally, when this javascript code is called on the client, it will call `V8Handler::Execute`, we will catch this and create an IPC message to let the Browser process know that an event has occured. In the Browser process, we catch this IPC message with the data and forward it to the listeners and do whatever you want with it.

Note that all javascript stuff happens in Renderer process so you need to do IPC messaging to communicate with it. Below I will list sample code step by step so you can inspect it and implement yourself.

Here is a sample [extension code](https://github.com/arkenthera/YumeEngine/blob/master/Engine/Assets/UI/Cef3D.js#L49). Attention to the `native` keyword, as it is [required](https://bitbucket.org/chromiumembedded/cef/wiki/JavaScriptIntegration.md#markdown-header-functions-and-extensions) by extensions. This code will be loaded in [renderer process](https://github.com/arkenthera/cef3d/blob/33980e114b18bd7736e70fa9e2c7d1ec27a3e2e2/CefSubProcess/Main.cc#L40) which then you will call `CefRegisterExtension` when Webkit is [initialized](https://github.com/arkenthera/cef3d/blob/33980e114b18bd7736e70fa9e2c7d1ec27a3e2e2/cef3d/Cef3DRendererApp.cc#L57).

When, for example, `native function SendDomEvent` is called on Javascript side, it will invoke `V8Handler::Execute` [here](https://github.com/arkenthera/cef3d/blob/33980e114b18bd7736e70fa9e2c7d1ec27a3e2e2/cef3d/Cef3DV8Handler.cc#L35) and we will create an IPC to let the browser know.

Then in browser process, we [catch](https://github.com/arkenthera/cef3d/blob/33980e114b18bd7736e70fa9e2c7d1ec27a3e2e2/cef3d/Cef3DBrowserDelegate.cc#L81) this IPC and after that its up to you to whatever you want with it.

Here is an example of turning a check box on and off, and letting the C++ side know about it.

```
Cef3D.SendDomEvent(this.onSendDomEvent,$(element.target).attr("id"),"InputChecked","" + $(element.target).is(':checked') + "");
```

This will then be catched [here](https://github.com/arkenthera/YumeEngine/blob/fe97366896e19fa5dd7ee17719bb3d24153f2efa/Engine/Source/Runtime/UI/YumeDebugOverlay.cc#L71).

I hope this helps!

**Comment by zbozo** [comment]
I am very grateful for the explanation and the codes shown
-- removed
oh, sorry, I got it fixed, I was running the wrong javascript, after hours I realized 👍 

I finally managed to make interaction

**Comment by 0x61726b** [comment]
I am not really sure what's happening there. Maybe check if `Cef3DRendererApp::OnContextReleased` is called or not to understand maybe something is causing that context to be invalid, thus crashing on `Enter()` ?

**Comment by 0x61726b** [comment]
Just saw your edit, nice!
